lib/content_negotiation_filter.js

```diff
@@ -61,6 +104,10 @@ module.exports = (hyper, req, next, options, specInfo) => {
                     mwUtil.extractHTMLProfileVersion(newRes.headers['content-type']);
                 if (semver.satisfies(newVersion, `^${requestedVersion}`)) {
                     return newRes;
+                } else if (semver.gt(newVersion, requestedVersion)) {
```

Comment:
Two lingering things from reading the above.

* On L68, the return `if  (!requestedVersion) {` means that it'll just return whatever is in storage, which is not quite the same thing as meaning it'll return the latest version.  Does that matter?

* Does the `no-cache` request on L101 store the response if the returned version is greater?

Reply:
> On L68, the return if (!requestedVersion) { means that it'll just return whatever is in storage, which is not quite the same thing as meaning it'll return the latest version. Does that matter?

If we did a `no-cache` for updating the version in that case it would be very prone to DDOS. If we hard-coded in RESTBase which version is currently the latest, that's really tight coupling. We're aiming to have all the clients provide the `Accept` header, so I consider lack of it a bug on the client side, so no guarantees for the buggy client.

> Does the no-cache request on L101 store the response if the returned version is greater?

It stores whatever Parsoid gives back. We aim to have latest parsoid version in storage, so why miss this opportunity to update

Reply:
> > Does that matter?
> If we did a no-cache for updating the version in that case it would be very prone to DDOS. 

Right, I was asking because our announcement tldr says,

> If omitted, they will always get the latest version of the HTML, regardless of any breaking changes it may contain.

which is incorrect.

> It stores whatever Parsoid gives back

Guess you need to be careful about not setting `Accept` when making parse requests then!!


Final question, what's the benefit of blanking the patch level (`requestedVersion = requestedVersion.replace(/\d+$/, '0');`)?  Clients will most often be behind on the patch level and benefit from the caret semantics.  But maybe a client really wants to ensure a bug fix that came in a patch update?

Reply:
> Guess you need to be careful about not setting Accept when making parse requests then!!

We totally don't: https://github.com/wikimedia/restbase/blob/master/v1/content.yaml#L223-L227

> Final question, what's the benefit of blanking the patch level (requestedVersion = requestedVersion.replace(/\d+$/, '0');)? Clients will most often be behind on the patch level and benefit from the caret semantics. But maybe a client really wants to ensure a bug fix that came in a patch update?

We actually are doing the same thing in Varnish to avoid too drastic cache fragmentation, so I've added the same logic here in order to make it work predictably with/without Varnish in front of it. After https://phabricator.wikimedia.org/T170606 is resolved (soon) and we accumulate some data we can see whether the cache fragmentation really be a problem for the clients or not and maybe reconsider the decision to ignore the semver patch version